lib/puppet/file_serving/http_metadata.rb

```diff
@@ -0,0 +1,29 @@
+require 'puppet/file_serving/metadata'
+
+# Simplified metadata representation, suitable for the information
+# that is available from HTTP headers.
+class Puppet::FileServing::HttpMetadata < Puppet::FileServing::Metadata
+
+  def initialize(http_response, path = '/dev/null')
+    super(path)
+
+    # ignore options that do not apply to HTTP metadata
+    @owner = @group = @mode = nil
+
+    if checksum = http_response['content-md5']
+      # convert base64 digest to hex
+      checksum = checksum.unpack("m0").first.unpack("H*").first
+      @checksum_type = 'md5'
+      @checksum = "{md5}#{checksum}"
+    elsif mtime = DateTime.httpdate(http_response['last-modified']).to_time
+      @checksum_type = 'mtime'
+      @checksum = "{mtime}#{mtime}"
+    else
+      raise PuppetError, "HTTP response contained no usable checksum equivalent"
+    end
+
+    @ftype = 'file'
```

Comment:
This was a question of semantics to me.

A HTTP source can only ever offer a flat file (unless we implement a spider in Puppet that can download directory trees), never a link or a directory.